src/tests/psyGen_test.py

```diff
@@ -810,3 +815,794 @@ def test_invalid_reprod_pad_size():
             "REPROD_PAD_SIZE in config.py should be a positive "
             "integer") in str(excinfo.value)
     config.REPROD_PAD_SIZE = keep
+
+
+def test_argument_dependent_arg():
+    '''Check that the dependent_arg method returns the appropriate boolean
+    value for arguments with combinations of read and write access'''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "4.5_multikernel_invokes.f90"),
+                           distributed_memory=False, api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3", distributed_memory=False).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    arg_f1_inc_1 = schedule.children[0].children[0].arguments.args[0]
+    arg_f1_inc_2 = schedule.children[2].children[0].arguments.args[0]
+    arg_f2_read_1 = schedule.children[0].children[0].arguments.args[2]
+    arg_f2_inc = schedule.children[1].children[0].arguments.args[0]
+    arg_f2_read_2 = schedule.children[2].children[0].arguments.args[1]
+    # different names returns False
+    assert not arg_f2_inc._dependent_arg(arg_f1_inc_1)
+    # same name both reads returns False
+    assert not arg_f2_read_1._dependent_arg(arg_f2_read_2)
+    # same name both incs (write to read) returns True
+    assert arg_f1_inc_2._dependent_arg(arg_f1_inc_1)
+    # read to write returns True
+    assert arg_f2_read_1._dependent_arg(arg_f2_inc)
+    # write to read returns True
+    assert arg_f2_inc._dependent_arg(arg_f2_read_1)
+    # same name both writes (the 4.5 example only uses inc) returns True
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH, "15.1_builtin_and_normal_kernel_invoke.f90"),
+        distributed_memory=False, api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3", distributed_memory=False).create(invoke_info)
+    invoke = psy.invokes.invoke_list[0]
+    schedule = invoke.schedule
+    arg_f1_write_1 = schedule.children[0].children[0].arguments.args[1]
+    arg_f1_write_2 = schedule.children[1].children[0].arguments.args[1]
+    assert arg_f1_write_1._dependent_arg(arg_f1_write_2)
```

Comment:
I had this as False for quite a while then realised that you should never have two writers following each other so thought it should probably raise an exception or warning. In the end I went with the simplest solution which was to allow it for the moment. Once you allow it then there is a dependence as by changing the order of writers you may get different results for any subsequent read. For example

a=1
a=2
b=a

The first and second writes to "a" depend on each other as we can not reorder them because b will get a different result.